File: kernel/src/main/kotlin/ninja/blacknet/transaction/TxData.kt
/*
 * Copyright (c) 2018-2024 Pavel Vasin
 *
 * Licensed under the Jelurida Public License version 1.1
 * for the Blacknet Public Blockchain Platform (the "License");
 * you may not use this file except in compliance with the License.
 * See the LICENSE.txt file at the top-level directory of this distribution.
 */

package ninja.blacknet.transaction

import ninja.blacknet.core.*
import ninja.blacknet.crypto.Hash

interface TxData {
    fun processCoinImpl(tx: Transaction, hash: Hash, dataIndex: Int, coinTx: CoinTx): Status

    fun processCoin(tx: Transaction, hash: Hash, coinTx: CoinTx): Status {
        val account = coinTx.getAccount(tx.from)
        if (account == null) {
            return Invalid("Sender account not found")
        }
        if (tx.seq != account.seq) {
            if (tx.seq.toUInt() < account.seq.toUInt()) {
                return AlreadyHave("sequence ${tx.seq} expected ${account.seq}")
            } else {
                require(tx.seq.toUInt() > account.seq.toUInt())
                return InFuture("sequence ${tx.seq} expected ${account.seq}")
            }
        }
        val status = account.credit(tx.fee)
        if (status != Accepted) {
            return notAccepted("Transaction fee", status)
        }
        account.seq += 1
        coinTx.setAccount(tx.from, account)
        return processCoinImpl(tx, hash, 0, coinTx)
    }
}

File: kernel/src/main/rust/transaction/txdata.rs
/*
 * Copyright (c) 2018-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use crate::blake2b::Hash;
use crate::error::{Error, Result};
use crate::transaction::{CoinTx, Transaction};
use alloc::format;

pub trait TxData {
    fn process_impl(
        &self,
        tx: Transaction,
        hash: Hash,
        data_index: u32,
        coin_tx: impl CoinTx,
    ) -> Result<()>;

    fn process(&self, tx: Transaction, hash: Hash, coin_tx: impl CoinTx) -> Result<()> {
        let mut account = coin_tx.get_account(tx.from())?;
        if tx.seq() != account.seq() {
            let msg = format!("sequence {} expected {}", tx.seq(), account.seq());
            if tx.seq() < account.seq() {
                return Err(Error::AlreadyHave(msg));
            } else {
                return Err(Error::InFuture(msg));
            }
        }
        account.credit(tx.fee())?;
        account.increment_seq();
        coin_tx.set_account(tx.from(), account);
        self.process_impl(tx, hash, 0, coin_tx)
    }
}
